File: app/utils/data_template_manager.py
from typing import Any
import json
from pathlib import Path

class DataTemplateManager:
    
    def __init__(self, json_filename:str) -> None:
        self.json_filename = json_filename
        self._template = self._load_json_templates()
    
    def _load_json_templat(self) -> dict:
        static_json_path = Path(__file__).parents[1] / "static" / "json" / self.json_filename
        
        if not static_json_path.exists():
            raise FileNotFoundError(f"JSON file {self.json_filename} not found.")
        
        with open(static_json_path, 'r') as file:
            return json.load(file)
    
    def set_value(self, key:str, value:Any) -> None:
        if key in self._template:
            self._template[key] = value
        else:
            raise KeyError(f"Key {key} not found in template {self.json_filename}.")
        
    def get_value(self, key: str) -> Any:
        if key in self._template:
            return self._template[key]
        else:
            raise KeyError(f"Key {key} not found in template {self.json_filename}.")
    
    def to_json(self) -> str:
        return json.dumps(self._templates)
    
    def from_json(self, json_string: str) -> None:
        try:
            data = json.loads(json_string)
            for key, value in data.items():
                if key in self._template:
                    self._template[key] = value
                else:
                    raise KeyError(f"Key {key} not found in template {self.json_filename}.")
        except json.JSONDecodeError:
            raise ValueError("Invalid JSON string provided.")

File: app/static/js/dataTemplateManager.js
class DataTemplateManager  {
    constructor(jsonFilename) {
        this.jsonFilename = jsonFilename;
        this._template = null;
    }

    async _loadJsonTemplate() {
        const response = await fetch(`../json/${this.jsonFilename}`);
        
        if (!response.ok) {
            throw new Error(`JSON file ${this.jsonFilename} not found.`);
        }
        
        this._template = await response.json();
    }

    async initialize() {
        await this._loadJsonTemplate();
    }

    setValue(key, value) {
        if (this._template && key in this._template) {
            this._template[key] = value;
        } else {
            throw new Error(`Key ${key} not found in template ${this.jsonFilename}.`);
        }
    }

    getValue(key) {
        if (this._template && key in this._template) {
            return this._template[key];
        } else {
            throw new Error(`Key ${key} not found in template ${this.jsonFilename}.`);
        }
    }

    toJson() {
        return JSON.stringify(this._template);
    }

    fromJson(jsonString) {
        try {
            const data = JSON.parse(jsonString);
            for (const key in data) {
                if (key in this._template) {
                    this._template[key] = data[key];
                } else {
                    throw new Error(`Key ${key} not found in template ${this.jsonFilename}.`);
                }
            }
        } catch (error) {
            throw new Error("Invalid JSON string provided.");
        }
    }
}
